sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -483,41 +533,107 @@ public void validate(PipelineOptions options) {
-      BoundedSource<TableRow> source;
-
-      if (getQuery() != null
-          && (!getQuery().isAccessible() || !Strings.isNullOrEmpty(getQuery().get()))) {
-        source =
-            BigQueryQuerySource.create(
-                stepUuid,
-                getQuery(),
-                getFlattenResults(),
-                getUseLegacySql(),
-                getBigQueryServices());
+      Pipeline p = input.getPipeline();
+      final PCollectionView<String> jobIdTokenView;
+      PCollection<String> jobIdTokenCollection = null;
+      PCollection<TableRow> rows;
+      if (!getUseNewSource()) {
+        // Create a singleton job ID token at construction time.
+        final String staticJobUuid = BigQueryHelpers.randomUUIDString();
+        jobIdTokenView = p
+            .apply("TriggerIdCreation", Create.of(staticJobUuid))
+            .apply("ViewId", View.<String>asSingleton());
+        // Apply the traditional Source model.
+        rows = p.apply(
+            org.apache.beam.sdk.io.Read.from(createSource(staticJobUuid)))
+            .setCoder(getDefaultOutputCoder());
       } else {
-        source =
-            BigQueryTableSource.create(
-                stepUuid,
-                getTableProvider(),
-                getBigQueryServices());
+        // Create a singleton job ID token at execution time.
+        jobIdTokenCollection = p
+            .apply("TriggerIdCreation", Create.of("ignored"))
+            .apply("CreateJobId", MapElements.via(
+                new SimpleFunction<String, String>() {
+                  @Override
+                  public String apply(String input) {
+                    return BigQueryHelpers.randomUUIDString();
+                  }
+                }));
+        jobIdTokenView = jobIdTokenCollection
+            .apply("ViewId", View.<String>asSingleton());
+
+        final TupleTag<String> filesTag =
+            new TupleTag<String>(){};
+        final TupleTag<String> tableSchemaTag =
+            new TupleTag<String>(){};
+        PCollectionTuple tuple = jobIdTokenCollection.apply("RunCreateJob", ParDo
+            .of(
+                new DoFn<String, String>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext c)
+                      throws Exception {
+                    String jobUuid = c.element();
+                    BigQuerySourceBase source = createSource(
+                        jobUuid);
+                    String schema = BigQueryHelpers.toJsonString(
+                        source.getSchema(c.getPipelineOptions()));
+                    c.output(tableSchemaTag, schema);
+                    List<ResourceId> files = source.extractFiles(c.getPipelineOptions());
+                    for (ResourceId file : files) {
+                      c.output(file.toString());
+                    }
+                  }
+                })
+            .withOutputTags(filesTag, TupleTagList.of(tableSchemaTag)));
```

Comment:
By coders, I meant:
- Remove `{}` at lines 565, 567
- Here, say: `tuple.get(filesTag).setCoder(StringUtf8Coder.of()); tuple.get(tableSchemaTag).setCoder(StringUtf8Coder.of());`